### ISSUE TITLE
Problems with the zero-shot eval pipeline (In sentence_zero_shot/compute_results.py) leading to incorrect evaluations

### ISSUE THREAD

**Issue by serdardoesml** [babylm/evaluation-pipeline-2025#34]
I found some hardware-dependent inconsistencies when evaluating BabyLM25 checkpoints. Our unstable amlm-hd-fail model that looked strong on EWoK/COMPS now scores ~50% (chance) once i fixed the following 3 issues. 

**1- Deterministic tie-break in minimal-pair ranking (main issue).**

```
stacked_probs = torch.stack(all_log_probs[temp], dim=1)
chosen_sentences = torch.max(stacked_probs, dim=1)[1].tolist()
```

torch.max(...)[1] always picks index 0 on ties. On 2-choice tasks (e.g., COMPS, BLiMP) this can inflate accuracy (up to almost %100 in some cases) if candidate 0 is the acceptable sentence.

**2- Eval math uses model dtype, making evals inconsistent on different hardware and amplifying the tie-break issue.**

F.log_softmax(logits / temp, ...) and reductions inherit the model’s output dtype. On A100 nodes this often means bf16/TF32 during eval, which increases near-ties/rounding ties and amplifies the tie-break issue. This causes divergences when running the evaluations on A100 vs CPU/V100.

**3- Length bias from summing token log-probs over the scored span.**

`phrase_log_probs = torch.sum(target_log_probs * sentence_dict[f"{prefix}_phrase_mask"].to(DEVICE), dim=1)`

Using the sum (instead of mean) favors shorter tokenizations in the scored region. This is not the main bug but it adds a systematic bias when the two candidates differ in scored length.

**Results after fixing**

Unstable models that previously looked strong on EWoK/COMPS now evaluate at ~50% (chance), matching expectations. (Before the fix, amlm-hd-fail evaluated at 88% running on v100 nodes, and 99.8% running on a100 nodes)

A properly trained model (amlm-hard-nhot) still shows small, sensible shifts on some benchmarks (table below).

<p style="margin: 0.0px 0.0px 0.0px 0.0px; font: 20.0px 'Helvetica Neue'"><b>Amlm-hard-nhot results:</b></p>
<p style="margin: 0.0px 0.0px 0.0px 0.0px; font: 13.0px 'Helvetica Neue'; min-height: 15.0px"><br></p>

  | Old Eval | Fixed Eval
-- | -- | --
Blimp | 65.6 | 64.7
Supplement | 54.8 | 58.0
EWoK | 49.7 | 50.02
Eye-tracking | 8.1 | 8.1
Reading | 3.3 | 3.3
Entity Tracking | 43.7 | 37.66
Adjective Nominalization | 64.0 | 50
Past tense | 1.0 | 0.06
COMPS | 51.3 | 51.3

<br class="Apple-interchange-newline">

I’m happy to open a PR with the exact fixes i used if that’s helpful. Note that my patches were quick/rough and targeted only the zero_shot full evaluations; they’ll likely need a light cleanup and wider testing to cover other evaluation modes.

**Comment by lgcharpe** [comment]
I think opening a PR would be nice, we can help with some of the cleanup to the rough versions

**Comment by jumelet** [comment]
Just out of curiosity, could you also run the eval without that length-bias fix? The first 2 points are indeed bugs that need fixing, the third is more a modelling decision than an actual bug.

**Comment by lgcharpe** [comment]
The third problem is a long standing problem and different papers (including the recent OLMES paper: https://arxiv.org/pdf/2406.08446) have looked into it, and it is not always obvious whether to do length normalization

**Comment by lgcharpe** [comment]
For the tie-break, we could also randomize the correct sentence position, this would make the tie-break result in 50% if only label 0 is chosen

**Comment by SerdarCS** [comment]
> Just out of curiosity, could you also run the eval without that length-bias fix? The first 2 points are indeed bugs that need fixing, the third is more a modelling decision than an actual bug.

I ran the evaluations again without the lenght-bias fix, and the results were exactly the same. I double checked to make sure i didnt make a mistake, it seems like the length bias does not really affect anything, probably all the datasets have similar or equal sentence lengths for both options.

<p style="margin: 0.0px 0.0px 0.0px 0.0px; font: 20.0px 'Helvetica Neue'"><b>Amlm-hard-nhot results:</b></p>
<p style="margin: 0.0px 0.0px 0.0px 0.0px; font: 13.0px 'Helvetica Neue'; min-height: 15.0px"><br></p>

  | Old Eval | Fixed Eval | Fixed Without Length Normalization
-- | -- | -- | --
Blimp | 65.6 | 64.7 | 64.7
Supplement | 54.8 | 58.0 | 58.0
EWoK | 49.7 | 50.02 | 50.02
Eye-tracking | 8.1 | 8.1 | 8.1
Reading | 3.3 | 3.3 | 3.3
Entity Tracking | 43.7 | 37.66 | 37.66
Adjective Nominalization | 64.0 | 50 | 50
Past tense | 1 | 6 | 6
COMPS | 51.3 | 51.3 | 51.3


<br class="Apple-interchange-newline">

**Comment by serdardoesml** [comment]
> I think opening a PR would be nice, we can help with some of the cleanup to the rough versions

Sure, i just need push access to the repo, right now i can not push any code.

**Comment by serdardoesml** [comment]
> Just out of curiosity, could you also run the eval without that length-bias fix? The first 2 points are indeed bugs that need fixing, the third is more a modelling decision than an actual bug.

I ran it again, it seems without the length bias fix, the first 2 points only affect the failed amlm-hd-fail model, making it evaluate as expected, but the amlm-hard-nhot model evaluates almost exactly the same as before, no difference to the pre-fix scores.

If not having length normalization is intended, then just fixing the first 2 issues should be enough to prevent inflated scores without causing changes in the scores for most models.
